compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/interop/JMethodResolver.java

```diff
@@ -138,29 +137,30 @@ JMethod resolve(JMethodRequest jMethodRequest) {
                 .collect(Collectors.toList());
     }
 
-    private List<JMethod> resolveByParamCount(List<JMethod> jMethods, int expectedCount, BType receiverType) {
-        return jMethods.stream().filter(jMethod -> {
-            int count = jMethod.getParamTypes().length;
-            if (count == expectedCount) {
+    private List<JMethod> resolveByParamCount(List<JMethod> jMethods, JMethodRequest jMethodRequest) {
+        return jMethods.stream()
+                .filter(jMethod -> hasEqualParamCounts(jMethodRequest, jMethod))
+                .collect(Collectors.toList());
+    }
+
+    private boolean hasEqualParamCounts(JMethodRequest jMethodRequest, JMethod jMethod) {
+        int expectedCount = getBFuncParamCount(jMethodRequest, jMethod);
+        int count = jMethod.getParamTypes().length;
+        if (count == expectedCount) {
+            return true;
+        } else if (count == expectedCount + 1) {
+            // This is for object interop functions when self is passed as a parameter
+            if (jMethodRequest.receiverType != null) {
+                jMethod.setReceiverType(jMethodRequest.receiverType);
                 return true;
-            } else {
-                boolean hasOneExtraParam = count == expectedCount + 1;
-                if (hasOneExtraParam) {
-                    boolean hasReceiver = receiverType != null;
-                    if (hasReceiver) {
-                        jMethod.setReceiverType(receiverType);
-                        return true;
-                    } else if (jMethod.isBalEnvAcceptingMethod()) {
-                        return true;
-                    }
-                }
+            } else if (jMethod.isStatic()) {
```

Comment:
do we have to check for static here?

Reply:
Yes, haven't tested with instance methods yet.